crates/red_knot_python_semantic/src/types/call/arguments.rs

```diff
@@ -11,16 +11,18 @@ impl<'a> CallArguments<'a> {
     /// Invoke a function with an optional extra synthetic argument (for a `self` or `cls`
     /// parameter) prepended to the front of this argument list. (If `bound_self` is none, the
     /// function is invoked with the unmodified argument list.)
-    pub(crate) fn with_self<F, R>(&mut self, bound_self: Option<Type<'_>>, f: F) -> R
+    pub(crate) fn with_self<F, R>(&self, bound_self: Option<Type<'_>>, f: F) -> R
     where
-        F: FnOnce(&mut Self) -> R,
+        F: FnOnce(&Self) -> R,
     {
+        let mut call_arguments = self.clone();
```

Comment:
We don't need to clone when there's no `bound_self`.  Opened #17793 for that, and to combine the `clone`/`push_front` into a single `collect` call, which should avoid a double-copy when there _is_ a `bound_self`.  (Or really, a copy + immediate move)